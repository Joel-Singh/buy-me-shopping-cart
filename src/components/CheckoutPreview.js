function CheckoutPreview({ cart }) {
  const cartAsJSX =
    cart.map((cartItem) =>
    {
      const {name, price, amountBought, image} = cartItem;
      return (
        <CheckoutPreviewItem
          name={name}
          price={price}
          amountBought={amountBought}
          image={image}
          key={name}
        />
      )
    });

  const totalCost = getOverallCost(cart);

  return (
    <div className="checkout-preview">
      <h2>Checkout</h2>
      <div data-testid="previewItemContainer">{cartAsJSX}</div>
      <div className="checkout-preview__total-cost">Total Cost = {totalCost}</div>
    </div>
  );
}

function CheckoutPreviewItem(props) {
  const { image, amountBought, price, name, removeItemCallback } = props;
  const totalPrice = amountBought * price;
  return (
    <div className="checkout-preview-item">
      <span className="checkout-preview-item__name">{name}</span>
      <img src={image} />
      <span className="checkout-preview-item__cart-amount">Amount in cart: {amountBought}</span>
      <span className="checkout-preview-item__total-price">
        Total Price: {totalPrice} ({amountBought} X {price})
      </span>
      <button type="button" onClick={removeItemCallback}>
        Remove Item
      </button>
    </div>
  );
}

function getOverallCost(cart) {
  const totalPriceOfEachItem = cart.map((cartItem) => {
    const { price, amountBought } = cartItem;
    return price * amountBought;
  });

  const overallCost = totalPriceOfEachItem.reduce((prev, curr) => {
    return prev + curr;
  }, 0);

  return overallCost;
}

export default CheckoutPreview;
