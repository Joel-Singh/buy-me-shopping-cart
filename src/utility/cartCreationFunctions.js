export function getInitialCart() {
  return (
    [
      createCartItem('Glasses', 12, 0, 'glasses placeholder'),
      createCartItem('Strand of my hair', 50, 0, 'Strand of my hair placeholder'),
      createCartItem('Programmer Bath Water in a Bottle', 30, 0, 'Programmer Bath Water in a Bottle placeholder'),
      createCartItem('A shirt with my face', 5, 0, 'A shirt with my face placeholder'),
      createCartItem('My empty, used toothpaste tube', 0, 30, 'My empty, used toothpaste tube placeholder'),
      createCartItem('My Used Razors', 30, 0, 'My Used Razors placeholder'),
      createCartItem('Donation', 500, 0, 'Donation placeholder'),
      createCartItem('Jared', 200, 0, 'Jared placeholder'),
      createCartItem(`Gum i've chewed`, 200, 0, `Gum i've chewed placeholder`),
      createCartItem('Lightly used pickle ;)', 200, 0, 'Lightly used pickle ;) placeholder'),
      createCartItem('My pet dog', 200, 0, 'My pet dog placeholder')
    ]
  )
}

export function createCartItem(name, price, amountBought, image) {
  return ({name, price, amountBought, image});
}
