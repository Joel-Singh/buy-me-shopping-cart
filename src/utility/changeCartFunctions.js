function addItemToCart(cart, itemName) {
  const clonedCart = JSON.parse(JSON.stringify(cart));
  const foundCartItem = findCartItem(clonedCart, itemName);
  ++foundCartItem.amountBought;
  return clonedCart;
}

function removeItemFromCart(cart, itemName) {
  const clonedCart = JSON.parse(JSON.stringify(cart));
  const foundCartItem = findCartItem(clonedCart, itemName);
  --foundCartItem.amountBought;
  return clonedCart;
}

function findCartItem(cart, itemName) {
  const foundCartItem =
    cart.find(
      (cartItem) =>
        cartItem.name === itemName
    );
  return foundCartItem;
}

export { addItemToCart, removeItemFromCart };
