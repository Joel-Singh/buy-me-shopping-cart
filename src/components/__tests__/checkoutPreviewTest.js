import CheckoutPreview from "../CheckoutPreview";
import { render, screen } from "@testing-library/react";
import { createCartItem } from './../../utility/cartCreationFunctions'
import userEvent from "@testing-library/user-event";

test("CheckoutPreviewItem renders a single preview item", () => {
  const cart = [
    createCartItem("Me Smiling", 32, 5, "placeholder"),
  ]

  render(<CheckoutPreview cart={cart} />)

  const previewItemContainer = screen.getByTestId("previewItemContainer")

  expect(previewItemContainer).toMatchSnapshot()
})

test("CheckoutPreviewItem renders multiple items", () => {
  const cart = [
    createCartItem("Me Smiling", 32, 5, "placeholder"),
    createCartItem("Me With a frown", 65, 2, "placeholder"),
    createCartItem("Me", 19, 4, "placeholder"),
  ]

  render(<CheckoutPreview cart={cart} />)

  const previewItemContainer = screen.getByTestId("previewItemContainer")

  expect(previewItemContainer).toMatchSnapshot()
})

test("CheckoutPreview renders total cost", () => {
  const cart = [
    createCartItem("Me Smiling", 32, 5, "placeholder"),
    createCartItem("Me With a frown", 65, 2, "placeholder"),
    createCartItem("Me", 19, 4, "placeholder"),
  ]

  render(<CheckoutPreview cart={cart} />)
  const totalHTMLContainer = screen.getByText("Total Cost", {exact: false})
  expect(totalHTMLContainer).toMatchSnapshot()
})

test("CheckoutPreviewItem doesn't render items not bought", () => {
  const cart = [
    createCartItem("Me Smiling", 32, 0, "placeholder"),
  ]

  render(<CheckoutPreview cart={cart} />)

  const previewItemContainer = screen.getByTestId("previewItemContainer")


  expect(previewItemContainer.innerHTML).toBe("")
})
