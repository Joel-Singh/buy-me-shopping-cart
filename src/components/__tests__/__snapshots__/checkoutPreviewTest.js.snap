// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutPreview renders total cost 1`] = `
<div
  class="checkout-preview__total-cost"
>
  Total Cost = 
  366
</div>
`;

exports[`CheckoutPreviewItem renders a single preview item 1`] = `
<div
  data-testid="previewItemContainer"
>
  <div
    class="checkout-preview-item"
  >
    <span
      class="checkout-preview-item__name"
    >
      Me Smiling
    </span>
    <img
      src="placeholder"
    />
    <span
      class="checkout-preview-item__cart-amount"
    >
      Amount in cart: 
      5
    </span>
    <span
      class="checkout-preview-item__total-price"
    >
      Total Price: 
      160
       (
      5
       X 
      32
      )
    </span>
    <button
      type="button"
    >
      Remove Item
    </button>
  </div>
</div>
`;

exports[`CheckoutPreviewItem renders multiple items 1`] = `
<div
  data-testid="previewItemContainer"
>
  <div
    class="checkout-preview-item"
  >
    <span
      class="checkout-preview-item__name"
    >
      Me Smiling
    </span>
    <img
      src="placeholder"
    />
    <span
      class="checkout-preview-item__cart-amount"
    >
      Amount in cart: 
      5
    </span>
    <span
      class="checkout-preview-item__total-price"
    >
      Total Price: 
      160
       (
      5
       X 
      32
      )
    </span>
    <button
      type="button"
    >
      Remove Item
    </button>
  </div>
  <div
    class="checkout-preview-item"
  >
    <span
      class="checkout-preview-item__name"
    >
      Me With a frown
    </span>
    <img
      src="placeholder"
    />
    <span
      class="checkout-preview-item__cart-amount"
    >
      Amount in cart: 
      2
    </span>
    <span
      class="checkout-preview-item__total-price"
    >
      Total Price: 
      130
       (
      2
       X 
      65
      )
    </span>
    <button
      type="button"
    >
      Remove Item
    </button>
  </div>
  <div
    class="checkout-preview-item"
  >
    <span
      class="checkout-preview-item__name"
    >
      Me
    </span>
    <img
      src="placeholder"
    />
    <span
      class="checkout-preview-item__cart-amount"
    >
      Amount in cart: 
      4
    </span>
    <span
      class="checkout-preview-item__total-price"
    >
      Total Price: 
      76
       (
      4
       X 
      19
      )
    </span>
    <button
      type="button"
    >
      Remove Item
    </button>
  </div>
</div>
`;
